dotnet-authserver/src/TeacherIdentity.AuthServer/Journeys/TrnTokenHelper.cs

```diff
@@ -0,0 +1,171 @@
+            {
+                if (existingValidUser.Trn is not null && existingValidUser.Trn != trnToken.Trn)
+                {
+                    // If we find a matching email in our records but TRN no match, ignore the token
+                    return;
+                }
+
+                authenticationState.OnSignedInUserProvided(existingValidUser);
+                if (existingValidUser.Trn is null)
+                {
+                    authenticationState.OnTrnTokenProvided(trnToken);
+                }
+                await authenticationState.SignIn(httpContext);
+            }
+            else
+            {
+                var users = await _userSearchService.FindUsers(trnToken.FirstName, trnToken.LastName, trnToken.DateOfBirth);
+                authenticationState.OnExistingAccountSearch(users.Length == 0 ? null : users[0]);
+                authenticationState.OnTrnTokenProvided(trnToken);
+            }
+        }
+    }
+
+    public async Task<User> ApplyTrnTokenToUser(Guid? userId, string trnTokenValue)
+    {
+        var user = await _dbContext.Users.SingleAsync(u => u.UserId == userId);
+        var trnToken = await _dbContext.TrnTokens.SingleAsync(t => t.TrnToken == trnTokenValue);
+
+        if (user.Trn is null)
+        {
+            user = await UpdateUserTrn(user, trnToken.Trn);
+        }
+
+        if (user.Trn == trnToken.Trn)
+        {
+            await InvalidateTrnToken(trnToken.TrnToken, user.UserId);
+        }
+
+        return user;
+    }
+
+    public async Task InvalidateTrnToken(string trnTokenValue, Guid userId)
+    {
+        await _dbContext.Database.ExecuteSqlInterpolatedAsync(
+            $"update trn_tokens set user_id = {userId} where trn_token = {trnTokenValue};");
+    }
+
+    private async Task<EnhancedTrnToken?> GetValidTrnTokenModel(string? trnTokenValue)
+    {
+        var trnToken = await _dbContext.TrnTokens.SingleOrDefaultAsync(t => t.TrnToken == trnTokenValue && t.ExpiresUtc > _clock.UtcNow);
+
+        if (trnToken is null || trnToken.UserId is not null)
+        {
+            return null;
+        }
+
+        var teacher = await _dqtApiClient.GetTeacherByTrn(trnToken.Trn);
+
+        if (teacher is null)
+        {
+            return null;
+        }
+
+        return new EnhancedTrnToken()
+        {
+            TrnToken = trnToken.TrnToken,
+            Trn = trnToken.Trn,
+            Email = trnToken.Email,
+            CreatedUtc = trnToken.CreatedUtc,
+            ExpiresUtc = trnToken.ExpiresUtc,
+            UserId = trnToken.UserId,
+            FirstName = teacher.FirstName,
+            MiddleName = teacher.MiddleName,
+            LastName = teacher.LastName,
+            DateOfBirth = teacher.DateOfBirth,
+        };
+    }
+
+    private async Task<User> UpdateUserTrn(User user, string trn)
+    {
+        user.Trn = trn;
+        user.TrnLookupStatus = TrnLookupStatus.Found;
+        user.CompletedTrnLookup = null;
```

Comment:
Not sure we want to necessarily `null` this out; if the user has gone through the lookup before (but not found a TRN) then I think we'd want to keep the original timestamp.